db/db_impl.cc

```diff
@@ -2929,10 +2936,30 @@ Status DBImpl::IngestExternalFile(
 
     // Make sure that bg cleanup wont delete the files that we are ingesting
     pending_output_elem = CaptureCurrentFileNumberInPendingOutputs();
+
+    // If crash happen after a hard link established, Recover function may
+    // reuse the file number that has already assigned to the internal file,
+    // and this will overwrite the external file. To protect the external
+    // file, we have to make sure the file number will never being reused.
+    next_file_number = versions_->FetchAddFileNumber(external_files.size());
+    auto cf_options = cfd->GetLatestMutableCFOptions();
+    status = versions_->LogAndApply(cfd, *cf_options, &dummy_edit, &mutex_,
+                                    directories_.GetDbDir());
+    if (status.ok()) {
+      InstallSuperVersionAndScheduleWork(cfd, &dummy_sv_ctx, *cf_options);
+    }
+  }
+  dummy_sv_ctx.Clean();
+  if (!status.ok()) {
+    return status;
+  }
+  if (next_file_number == 0) {
```

Comment:
I think fetch_add will never return 0 because VersionSet::next_file_number_ is initialized to 2 at the very beginning. If that happen, something must be wrong. Maybe I add an assert?

Reply:
I don't think we need this check because putting this check or an assertion seems overly defensive. Since the C++ standard and the variable's initial value guarantees that it will never be 0, let's just remove this check.

Reply:
Good idea